feat: `createRequire` option and disable esm for jest

clarkdo opened the pull request:
<!--- Provide a general summary of your changes in the title above -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (a non-breaking change which fixes an issue)
- [ ] New feature (a non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)


## Description
<!--- Describe your changes in detail -->
<!--- Why is this change required? What problem does it solve? -->
<!--- If it resolves an open issue, please link to the issue here. For example "Resolves: #1337" -->

### esm disabled by default

As [esm](https://github.com/standard-things/esm) package has been incompatible since [jest 25](https://jestjs.io/blog/2020/01/21/jest-25), this PR disables esm for jest environments by default to prevent regression happening when testing nuxt with jest. You may need to make changes to your project after this PR:

jest config:

```js
    "transform": {
      "^.+\\.js$": "babel-jest"
    },
```

babel config:

```js
module.exports = function (api) {
  if (api.env('test')) {
    return {
      plugins: ['@babel/plugin-transform-modules-commonjs']
    }
  }
  return {}
}
```

### `createRequire` option

Meanwhile, we have started working on [`jiti`](https://github.com/nuxt-contrib/jiti) project for stable esm/typescript runtime support. You can try it by setting `createRequire: 'jiti'` inside `nuxt.config`.

Also, it is possible to use this option to keep using `esm` with jest < 25 or provide a custom factory function. Setting value to false or unsupported value, makes nuxt using native require (via [create-require](https://github.com/nuxt-contrib/create-require))

## Checklist:
<!--- Put an `x` in all the boxes that apply. -->
<!--- If your change requires a documentation PR, please link it appropriately -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly. (PR: #)
- [ ] I have added tests to cover my changes (if not applicable, please state why)
- [ ] All new and existing tests are passing.

clarkdo committed:
fix(test): not use esm in jest

clarkdo committed:
Merge branch 'dev' into esm-jest

clarkdo commented:
@pi0 Feel free to merge or close if we're gonna remove esm.

pi0 commented:
@clarkdo PR is good we should remove esm more carefully. I was just thinking to provide an option instead of disabling. Can make changes in same PR :)

pi0 committed:
feat: options.createRequire

pi0 committed:
feat: jest/esm warning

pi0 committed:
fix resolver: name conflict

pi0 committed:
Merge branch 'dev' into esm-jest

Commit:
upadate load.js

Atinux commented on packages/config/src/load.js:
```diff
@@ -52,9 +53,8 @@ export async function loadNuxtConfig ({
-
-    options = esm(module)(configFile) || {}
-
+    const requirModule = createRequire(module)
```
```suggestion
    const requireModule = createRequire(module)
```

pi0 replied:
@Atinux PR is still WIP pushing fixes in few minutes :)

Atinux commented on packages/config/src/load.js:
```diff
@@ -52,9 +53,8 @@ export async function loadNuxtConfig ({
-    options = esm(module)(configFile) || {}
-
+    const requirModule = createRequire(module)
+    options = requirModule(configFile) || {}
```
```suggestion
    options = requireModule(configFile) || {}
```

Commit:
update resolver.js

Commit:
update resolver.test

Commit:
use create-require by default

- we already have this dependency
- jest mocks confirmed to working
- will help for pnp migration

Commit:
test: update snapshots

pi0 committed:
Merge branch 'dev' into esm-jest

Commit:
test: fix vue-app creareRequire

Commit:
disable for jest by default

pi0 committed:
Merge branch 'dev' into esm-jest

Commit:
update yarn.lock

Commit:
try to fix azure issue

Commit:
update options.test
